File: src-tauri/src/models/project.rs
//! @module models/project
//! @description Data models for projects, detection results, and health scores
//!
//! PURPOSE:
//! - Define the Project struct for database and IPC
//! - Define HealthScore and component breakdown types
//! - Define DetectionResult with full tech stack detection
//! - Define ProjectSetup for onboarding configuration
//!
//! DEPENDENCIES:
//! - serde - Serialization for Tauri IPC
//! - chrono - Timestamp handling
//!
//! EXPORTS:
//! - Project - Core project metadata stored in database
//! - HealthScore - Overall project health with component breakdown
//! - HealthComponents - Individual health component scores
//! - QuickWin - Prioritized improvement suggestion
//! - DetectionResult - Full auto-detection output from project scanning
//! - DetectedValue - A detected value with confidence and source
//! - ProjectSetup - Configuration collected during onboarding
//!
//! PATTERNS:
//! - All structs derive Clone, Debug, Serialize, Deserialize
//! - IDs use UUID v4 strings
//! - Timestamps use chrono::DateTime<Utc>
//! - Detection confidence: "high" (config files), "medium" (deps), "low" (extensions)
//!
//! CLAUDE NOTES:
//! - Keep in sync with TypeScript types in src/types/project.ts
//! - Health score range is always 0-100
//! - DetectionResult expanded in Phase 2 to include database, testing, styling

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    pub id: String,
    pub name: String,
    pub path: String,
    pub description: String,
    pub project_type: String,
    pub language: String,
    pub framework: Option<String>,
    pub database: Option<String>,
    pub testing: Option<String>,
    pub styling: Option<String>,
    pub health_score: u32,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HealthScore {
    pub total: u32,
    pub components: HealthComponents,
    pub quick_wins: Vec<QuickWin>,
    pub context_rot_risk: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HealthComponents {
    pub claude_md: u32,
    pub module_docs: u32,
    pub freshness: u32,
    pub skills: u32,
    pub context: u32,
    pub enforcement: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QuickWin {
    pub title: String,
    pub description: String,
    pub impact: u32,
    pub effort: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DetectionResult {
    pub confidence: String,
    pub language: Option<DetectedValue>,
    pub framework: Option<DetectedValue>,
    pub database: Option<DetectedValue>,
    pub testing: Option<DetectedValue>,
    pub styling: Option<DetectedValue>,
    pub project_name: Option<String>,
    pub project_type: Option<String>,
    pub file_count: u32,
    pub has_existing_claude_md: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DetectedValue {
    pub value: String,
    pub confidence: f64,
    pub source: String,
}

/// Configuration collected during onboarding wizard
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSetup {
    pub path: String,
    pub name: String,
    pub description: String,
    pub project_type: String,
    pub language: String,
    pub framework: Option<String>,
    pub database: Option<String>,
    pub testing: Option<String>,
    pub styling: Option<String>,
    pub goals: Vec<String>,
    pub generate_module_docs: bool,
    pub setup_enforcement: bool,
}

File: src/types/project.ts
/**
 * @module types/project
 * @description TypeScript type definitions for projects, detection, and onboarding
 *
 * PURPOSE:
 * - Define Project interface matching Rust model
 * - Define DetectionResult for project scanning
 * - Define ProjectSetup for onboarding wizard data
 * - Define StackExtras for additional services (auth, hosting, payments, etc.)
 * - Define project-related option constants
 *
 * EXPORTS:
 * - StackExtras - Additional services configuration (auth, hosting, payments, etc.)
 * - Project - Core project metadata
 * - DetectionResult - Auto-detection output from project scanning
 * - DetectedValue - A detected value with confidence level
 * - ClaudeMdInfo - Metadata about a CLAUDE.md file (exists, content, tokens)
 * - ProjectSetup - Configuration collected during onboarding
 * - LANGUAGES, FRAMEWORKS, DATABASES, etc. - Option lists for dropdowns
 * - AUTH_OPTIONS, HOSTING_OPTIONS, PAYMENTS_OPTIONS, MONITORING_OPTIONS, EMAIL_OPTIONS - Stack extras options
 *
 * PATTERNS:
 * - Types mirror Rust structs in models/project.rs
 * - Use camelCase (TypeScript convention), Rust uses snake_case
 *
 * CLAUDE NOTES:
 * - Keep in sync with Rust models in src-tauri/src/models/project.rs
 * - Tauri IPC automatically converts snake_case to camelCase
 * - ProjectSetup expanded in Phase 2 with full onboarding fields
 * - StackExtras fields are all optional (user may not select any)
 */

/**
 * Stack extras for additional services (auth, hosting, payments, etc.)
 */
export interface StackExtras {
  auth?: string;
  hosting?: string;
  payments?: string;
  monitoring?: string;
  email?: string;
  cache?: string;
}

export interface Project {
  id: string;
  name: string;
  path: string;
  description: string;
  projectType: string;
  language: string;
  framework: string | null;
  database: string | null;
  testing: string | null;
  styling: string | null;
  stackExtras: StackExtras | null;
  healthScore: number;
  createdAt: string;
}

export interface DetectionResult {
  confidence: "high" | "medium" | "low" | "none";
  language: DetectedValue | null;
  framework: DetectedValue | null;
  database: DetectedValue | null;
  testing: DetectedValue | null;
  styling: DetectedValue | null;
  projectName: string | null;
  projectType: string | null;
  fileCount: number;
  hasExistingClaudeMd: boolean;
}

export interface DetectedValue {
  value: string;
  confidence: number;
  source: string;
}

export interface ClaudeMdInfo {
  exists: boolean;
  content: string;
  tokenEstimate: number;
  path: string;
}

export interface ProjectSetup {
  path: string;
  name: string;
  description: string;
  projectType: string;
  language: string;
  framework: string | null;
  database: string | null;
  testing: string | null;
  styling: string | null;
  stackExtras: StackExtras | null;
  goals: string[];
  generateModuleDocs: boolean;
  setupEnforcement: boolean;
}

export const LANGUAGES = [
  "TypeScript",
  "JavaScript",
  "Python",
  "Rust",
  "Go",
  "Dart",
  "Java",
  "Kotlin",
  "Swift",
  "Ruby",
  "PHP",
  "C#",
  "C++",
  "C",
] as const;

export const PROJECT_TYPES = [
  "Web App",
  "API",
  "Mobile",
  "Desktop",
  "CLI",
  "Library",
  "Monorepo",
  "Extension",
] as const;

export const FRAMEWORKS: Record<string, string[]> = {
  TypeScript: [
    "React",
    "Next.js",
    "Vue",
    "Nuxt",
    "Angular",
    "Svelte",
    "SolidJS",
    "Express",
    "Fastify",
    "NestJS",
    "Hono",
    "Tauri",
    "Electron",
    "Chrome Extension",
  ],
  JavaScript: [
    "React",
    "Next.js",
    "Vue",
    "Nuxt",
    "Angular",
    "Svelte",
    "Express",
    "Fastify",
    "Electron",
    "Chrome Extension",
  ],
  Python: ["Django", "FastAPI", "Flask", "Starlette", "Tornado"],
  Rust: ["Tauri", "Actix Web", "Axum", "Rocket", "Warp", "Leptos", "Yew", "Dioxus"],
  Go: ["Gin", "Fiber", "Echo", "Gorilla Mux"],
  Dart: ["Flutter"],
  Java: ["Spring Boot", "Quarkus", "Micronaut"],
  Kotlin: ["Ktor", "Spring Boot"],
  Swift: ["SwiftUI", "Vapor"],
  Ruby: ["Rails", "Sinatra"],
  PHP: ["Laravel", "Symfony"],
};

export const DATABASES = [
  "PostgreSQL",
  "MySQL",
  "SQLite",
  "MongoDB",
  "Redis",
  "Supabase",
  "Firebase",
  "DynamoDB",
  "Pinecone",
  "Weaviate",
  "Qdrant",
  "Chroma",
  "Milvus",
  "PlanetScale",
  "CockroachDB",
  "Neon",
  "Turso",
] as const;

export const TESTING_FRAMEWORKS: Record<string, string[]> = {
  TypeScript: ["Vitest", "Jest", "Testing Library", "Playwright", "Cypress", "Mocha"],
  JavaScript: ["Jest", "Vitest", "Mocha", "Cypress", "Playwright"],
  Python: ["pytest", "unittest", "nose2"],
  Rust: ["cargo test", "insta (snapshot)"],
  Go: ["go test", "testify"],
};

export const STYLING_OPTIONS = [
  "Tailwind CSS",
  "CSS Modules",
  "Styled Components",
  "Emotion",
  "Sass/SCSS",
  "Material UI",
  "Chakra UI",
  "Bootstrap",
  "Less",
] as const;

export const GOALS = [
  { id: "features", label: "Writing new features faster", skill: "generators" },
  { id: "tests", label: "Writing tests", skill: "test-agent" },
  { id: "reviews", label: "Code reviews", skill: "code-reviewer" },
  { id: "refactoring", label: "Refactoring", skill: "refactor-agent" },
  { id: "debugging", label: "Debugging", skill: "debug-agent" },
  { id: "documentation", label: "Documentation", skill: "docs-agent" },
] as const;

export const AUTH_OPTIONS = [
  "Auth0",
  "Clerk",
  "NextAuth.js",
  "Supabase Auth",
  "Firebase Auth",
  "Custom JWT",
  "Resend",
] as const;

export const HOSTING_OPTIONS = [
  "Vercel",
  "Railway",
  "Render",
  "AWS",
  "Fly.io",
  "Netlify",
  "Cloudflare",
] as const;

export const PAYMENTS_OPTIONS = [
  "Stripe",
  "LemonSqueezy",
  "Paddle",
] as const;

export const MONITORING_OPTIONS = [
  "Sentry",
  "PostHog",
  "Datadog",
  "LogRocket",
] as const;

export const EMAIL_OPTIONS = [
  "SendGrid",
  "Postmark",
  "AWS SES",
] as const;
